audio/drivers/coreaudio.cpp

```diff
@@ -0,0 +1,384 @@
+        *audioUnit = NULL;
+        *audioDevice = kAudioDeviceUnknown;
+        return false;
+    }
+    
+    status = AudioUnitSetProperty(*audioUnit, kAudioOutputUnitProperty_CurrentDevice, kAudioUnitScope_Global, 0, audioDevice, sizeof(AudioDeviceID));
+    if(status) {
+        qDebug("Failed to set current device.");
+        AudioComponentInstanceDispose(*audioUnit);
+        *audioUnit = NULL;
+        *audioDevice = kAudioDeviceUnknown;
+        return false;
+    }
+    
+    AudioStreamBasicDescription desiredFormat;
+    memset(&desiredFormat, 0, sizeof(AudioStreamBasicDescription));
+    desiredFormat.mFormatID = kAudioFormatLinearPCM;
+    desiredFormat.mFormatFlags = kAudioFormatFlagsNativeFloatPacked;
+    desiredFormat.mFramesPerPacket = 1;
+    desiredFormat.mBitsPerChannel = sizeof(float) * 8;
+    
+    UInt32 converterQuality = kAudioConverterQuality_Max;
+    status = AudioUnitSetProperty(*audioUnit, kAudioUnitProperty_RenderQuality, kAudioUnitScope_Global, 0, &converterQuality, sizeof(converterQuality));
+    if(status) {
+        qDebug("Failed to set render quality");
+        goto error;
+    }
+    desiredFormat.mSampleRate = getSampleRate(*audioDevice);
+    desiredFormat.mBytesPerPacket = sizeof(float) * 2;
+    desiredFormat.mBytesPerFrame = sizeof(float) * 2;
+    desiredFormat.mChannelsPerFrame = 2;
+    status = AudioUnitSetProperty(*audioUnit, kAudioUnitProperty_StreamFormat, kAudioUnitScope_Input, 0, &desiredFormat, sizeof(desiredFormat));
+    if(status) {
+        qDebug("Failed to set stream format");
+        goto error;
+    }
+    
+    AURenderCallbackStruct renderCallbackData;
+    renderCallbackData.inputProc = CoreAudioIOProc;
+    renderCallbackData.inputProcRefCon = ca;
+    
+    status = AudioUnitSetProperty(*audioUnit, kAudioUnitProperty_SetRenderCallback, kAudioUnitScope_Output, 0, &renderCallbackData, sizeof(renderCallbackData));
+    if(status) {
+        qDebug("Failed to set render callback");
+        goto error;
+    }
+    
+    AudioUnitInitialize(*audioUnit);
+    
+    return true;
+   error:
+    if(*audioUnit) {
+        AudioComponentInstanceDispose(*audioUnit);
+        *audioUnit = NULL;
+    }
+    *audioDevice = kAudioDeviceUnknown;
+    return false;
+}
+
+static OSStatus CoreAudioIOProc( void *inRefCon,
+                               AudioUnitRenderActionFlags *ioActionFlags,
+                               const AudioTimeStamp *inTimeStamp,
+                               UInt32 inBusNumber,
+                               UInt32 inNumberFrames,
+                               AudioBufferList *ioData)
+{
+      seq->setInitialMillisecondTimestampWithLatency();
+
+      long bytesPerFrame = sizeof(float) * ioData->mBuffers[0].mNumberChannels;
+
+      float* buffer = static_cast<float*>(ioData->mBuffers[0].mData);
+      UInt64 frames = ioData->mBuffers[0].mDataByteSize / bytesPerFrame;
+
+      seq->process((unsigned)frames, buffer);
```

Comment:
You might want to do
```
      if (seq->driver())
            seq->process((unsigned)frames, buffer);
```
in order to prevent the crash at program exit.